### PR TITLE
CFE-2283: Add email filtering feature for cf-execd.

### PR DESCRIPTION
**Pull request by kacf**
Also add some tests for cf-execd, where we mock one function which is
the connection to the SMTP server. Instead we connect to a local pipe
and feed it data.

Changelog: New feature: Emails sent by cf-execd can be filtered to get
rid of emails for unwanted log messages. The attributes
`mailfilter_include` and `mailfilter_exclude` in `body executor
control` control what to include. See documentation for `cf-execd` for
more information.

### CODE REVIEW

**Comment by kacf** [comment]
Note, this PR contains within it #2584.

**Comment by kacf** [comment]
trigger build

**Comment by kacf** [comment]
trigger build

**Comment by kacf** [comment]
trigger build

**Comment by kacf** [comment]
Now green in Jenkins!

**Comment by kacf** [comment]
@pasinskim: I would appreciate a review of this if you have the time!

**Comment by kacf** [comment]
trigger build

**Comment by kacf** [comment]
trigger build

File: cf-execd/cf-execd-runner.c [hunk 1]
```diff
@@ -713,30 +735,33 @@ static void MailResult(const ExecConfig *config, const char *file)
...
         {
-            break;
+            continue;
         }
-
-        Log(LOG_LEVEL_DEBUG, "Mail report: %s", vbuff);
-
-        if (strlen(vbuff) > 0)
+        if (send(sd, line, read, 0) == -1 ||
+            send(sd, "\r\n", 2, 0) == -1)
```
Comment (pasinskim) [inline cf-execd/cf-execd-runner.c]: I can see that previously we've been sending just '\n'.

Comment (kacf) [inline cf-execd/cf-execd-runner.c]: Indeed, but this was a bug. The RFC clearly states that lines should be separated by CRLF.

File: cf-execd/cf-execd-runner.h [hunk 2]
```diff
@@ -29,5 +29,7 @@
 
 void LocalExec(const ExecConfig *config);
 
+extern int ConnectToSmtpSocket(const ExecConfig *config);
```
Comment (pasinskim) [inline cf-execd/cf-execd-runner.h]: Do we need ```extern``` here?

Comment (kacf) [inline cf-execd/cf-execd-runner.h]: Oh you're right, we don't. Part of an earlier experiment. Will remove it!

File: cf-execd/exec-config.c [hunk 3]
```diff
@@ -51,6 +51,59 @@ static char *GetIpAddresses(const EvalContext *ctx)
     return StringWriterClose(ipbuf);
 }
 
+static void RegexFree(void *ptr)
+{
+    pcre_free(ptr);
+}
+
+static void MailFilterFill(const void *input, bool rlist,
```
Comment (pasinskim) [inline cf-execd/exec-config.c]: This is rather matter of taste, but I would add two functions that you can call from inside ```FillList``` and ```FillSeq```. Having this you will avoid checking ```rlist``` condition over and over again.

Comment (kacf) [inline cf-execd/exec-config.c]: Yeah, this part is rather ugly. I'll see if I can restructure it a bit.

File: tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    { \
+        Log(LOG_LEVEL_ERR, "Failed to read SMTP response in %s:%i", __FILE__, __LINE__); \
+        return NULL; \
+    } \
+    else \
+    { \
+        fwrite(X, rcvd, 1, stdout); \
+    }
+
+    char *sendbuf;
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c]: Looks like ```sendbuf``` is ```const char *```.

File: tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    char recvbuf[CF_BUFSIZE];
+    int rcvd;
+    sendbuf = "220 test.com\r\n";
+    COND_SEND(sendbuf);
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c]: Why you need a ```sendbuff``` at all? Shouldn't ```COND_SEND("220 test.com\r\n")``` be enough?

File: tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+int ConnectToSmtpSocket(ARG_UNUSED const ExecConfig *config)
+{
+    int sockets[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) < 0)
+    {
+        return -1;
+    }
+
+    int *thread_socket = xmalloc(sizeof(int));
+    *thread_socket = sockets[1];
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c]: Why we can not simply pass ```sockets + 1``` or ```&sockets[1]``` as a parameter to ```pthread_create``` below?

File: tests/acceptance/25_cf-execd/mail_exclude_filter.cf [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# Tests that exclude mail filters, causes everything but those lines to be
+# mailed by cf-execd.
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/mail_exclude_filter.cf]: The description is hard to understand. Maybe something like "Tests that exclude mail filter causes excluded lines are not mailed by cf-execd.".

File: tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c (hunk 5 above)
Comment (kacf) [inline tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c]: Yep, it is, will fix!

File: tests/acceptance/25_cf-execd/mail_include_and_exclude_filters.cf [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                            "WrongString3",
+                          };
+      "includes" slist => {
+                            "R: CorrectString1", ".*CorrectString[2]",
+                            "R: WrongString1", ".*WrongString[2]"
+                          };
+      "excludes" slist => {
+                            "R: WrongString1", ".*WrongString[2]",
+                            # Tests anchoring.
+                            "orrectString1"
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/mail_include_and_exclude_filters.cf]: Is "orrectString1" expected here?

File: tests/acceptance/25_cf-execd/mail_include_filter.cf [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# Tests that include mail filters, causes only those lines to be mailed by
+# cf-execd.
```
Comment (pasinskim) [inline tests/acceptance/25_cf-execd/mail_include_filter.cf]: Maybe " Tests that include mail filter causes only matched lines are mailed by cf-execd."?

**Comment by pasinskim** [comment]
Looks nice expect some rather minor issues. The change is big though so was hard to check deeply. Definitely you've put lot of effort in adding extensive tests which is great!!!

File: tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c (hunk 6 above)
Comment (kacf) [inline tests/acceptance/25_cf-execd/cf-execd-rpl-functions.c]: We cannot pass `sockets + 1` or `&sockets[1]` because the variable is stack allocated, and this stack frame gets destroyed after we have returned from this function. We also cannot pass the int cast as a pointer because `sizeof(int) != sizeof(int*)` on 64-bit platforms. They are 4 and 8, respectively.

File: tests/acceptance/25_cf-execd/mail_exclude_filter.cf (hunk 7 above)
Comment (kacf) [inline tests/acceptance/25_cf-execd/mail_exclude_filter.cf]: Ok, will fix!

File: tests/acceptance/25_cf-execd/mail_include_filter.cf (hunk 9 above)
Comment (kacf) [inline tests/acceptance/25_cf-execd/mail_include_filter.cf]: Ok, will fix!

File: tests/acceptance/25_cf-execd/mail_include_and_exclude_filters.cf (hunk 8 above)
Comment (kacf) [inline tests/acceptance/25_cf-execd/mail_include_and_exclude_filters.cf]: Yes, we are testing that a partial match does not match. I can improve the comment a bit to clarify this.

File: cf-execd/exec-config.c (hunk 3 above)
Comment (kacf) [inline cf-execd/exec-config.c]: I split this into three functions, one for each of the `Seq` and `Rlist` parts, and one common one.

**Comment by kacf** [comment]
Issues fixed and branch updated. I will run this once more through Jenkins.

**Comment by kacf** [comment]
Thanks for the review, @pasinskim!

**Comment by kacf** [comment]
trigger build

**Comment by jimis** [comment]
Dammit @kacf, the effort you've put in testing in this pull request is really scary... :-)

**Comment by kacf** [comment]
Hehe, still, I think it would have taken me even longer to test this manually. It was quite tricky to get it right, and the test helped me a lot (I was plagued by a single inverted boolean for a long time during the implementation).

File: tests/acceptance/Makefile.am [hunk 10]
```diff
@@ -21,6 +21,9 @@
 # (COSL) may apply to this file if you as a licensee so wish it. See
 # included file COSL.txt.
 #
+
+SUBDIRS = 25_cf-execd
```
Comment (olehermanse) [inline tests/acceptance/Makefile.am]: @kacf Why is this needed for cf-execd but no other acceptance tests?

Comment (kacf) [inline tests/acceptance/Makefile.am]: Because we build the `cf-execd-test` binary there (there is a `Makefile.am` added there in this PR as well). The other tests don't build anything, and hence don't need this.

Comment (olehermanse) [inline tests/acceptance/Makefile.am]: Ah, I think I saw that. Nice.
